src/main/java/wolox/training/services/implement/OpenLibraryServiceImp.java

```diff
@@ -0,0 +1,103 @@
+import wolox.training.services.OpenLibraryService;
+
+@Service
+public class OpenLibraryServiceImp implements OpenLibraryService {
+
+    private static final String URL = "https://openlibrary.org/api/books";
+    private static final String PARAM_STRING = "%1$s?bibkeys=ISBN:%2$s%3$s";
+    private static final String JSON_FORMAT_STRING = "&format=json&jscmd=data";
+    private static final String BOOK_NOT_FOUND_MSG = "Book not found";
+    private static final String IMAGE_MSG = "Has no image";
+    private static final String ATTRIBUTE_CONFLICT_MSG = "An attribute has a conflict";
+
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    @Autowired
+    private BookRepository bookRepository;
+
+    @Autowired
+    private ModelMapper modMapper;
+
+
+    private BookDTO bookInfo(String isbn) {
+        String url = String.format(PARAM_STRING,URL,isbn,JSON_FORMAT_STRING);
+        ResponseEntity<Object> response = restTemplate.getForEntity(url , Object.class);
+        LinkedHashMap body = (LinkedHashMap) response.getBody();
+        if (body.size() == 0){
+            throw new BookNotFoundException(BOOK_NOT_FOUND_MSG);
+        }
+        BookDTO bookInfo = mapBookInfo((LinkedHashMap) body.get("ISBN:" + isbn), isbn);
+        return bookInfo;
+
+
+    }
+
+    private BookDTO mapBookInfo(LinkedHashMap data, String isbn) {
+        try {
+            BookDTO bookInfo = new BookDTO();
+            List<String> authorNames = new ArrayList<>();
+            List<String> publisherNames = new ArrayList<>();
+            bookInfo.setIsbn(isbn);
+            bookInfo.setTitle(data.get("title").toString());
+            bookInfo.setSubtitle(data.get("subtitle").toString());
+            bookInfo.setYear(data.get("publish_date").toString());
+            bookInfo.setPages(
+                    (Integer) data.get("number_of_pages"));
+            List<LinkedHashMap> publishers = new ArrayList<>(
+                    (Collection<? extends LinkedHashMap>) data.get("publishers")
+            );
+
+            for (LinkedHashMap p : publishers) {
+                publisherNames.add(p.get("name").toString());
+            }
+            bookInfo.setPublishers(publisherNames);
+
+            List<LinkedHashMap> authors = new ArrayList<>(
+                    (Collection<? extends LinkedHashMap>) data.get("authors")
+            );
+
+            for (LinkedHashMap a : authors) {
+                authorNames.add(a.get("name").toString());
+            }
+            bookInfo.setAuthors(authorNames);
+            return bookInfo;
+        } catch (RuntimeException ex) {
+            throw new BookAttributeConflict(ATTRIBUTE_CONFLICT_MSG);
+        }
+    }
+
+    @Override
+    public Book getBookByISBN(String isbn) {
+        try {
+            BookDTO isBookApi = null;
+            isBookApi = bookInfo(isbn);
+            Book bookToSave = modMapper.map(isBookApi, Book.class);
+            bookToSave.setAuthor(isBookApi.getAuthors().get(0));
```

Comment:
you could do a buil in Book Class, where recieve a BookDTO. Otherwise It would be to create an adapter class that is in charge of adapting all the objects that arrive (external objects) to an object of the system